pkg/proxy/proxy_community_test.go

```diff
@@ -0,0 +1,374 @@
+				},
+				ExpectedProxyResponseStatus: http.StatusOK,
+				RequestPath:                 podsRequestURL("smith-community"),
+				ExpectedResponse:            httpTestServerResponse,
+			},
+			// Given smith owns a workspace named smith-community
+			// And   smith-community is publicly visible (shared with PublicViewer)
+			// And   a user named alice exists
+			// And   smith's smith-community is not directly shared with alice
+			// When  alice requests the list of pods in workspace smith-community
+			// Then  the request is forwarded from the proxy
+			// And   the request impersonates the PublicViewer
+			// And   the request's X-SSO-User Header is set to alice's ID
+			// And   the request is successful
+			"plain http actual request as community user": {
+				ProxyRequestMethod:  "GET",
+				ProxyRequestHeaders: map[string][]string{"Authorization": {"Bearer " + s.token(alice)}},
+				ExpectedAPIServerRequestHeaders: map[string][]string{
+					"Authorization":    {"Bearer clusterSAToken"},
+					"Impersonate-User": {toolchainv1alpha1.KubesawAuthenticatedUsername},
+					"X-SSO-User":       {"username-" + alice.String()},
+				},
+				ExpectedProxyResponseStatus: http.StatusOK,
+				RequestPath:                 podsRequestURL("smith-community"),
+				ExpectedResponse:            httpTestServerResponse,
+			},
+			// Given user alice exists
+			// And   alice owns a private workspace
+			// When  smith requests the list of pods in alice's workspace
+			// Then  the request is forwarded from the proxy
+			// And   the request impersonates smith
+			// And   the request's X-SSO-User Header is set to smith's ID
+			// And   the request is NOT successful
+			"plain http actual request as non-owner to private workspace": {
+				ProxyRequestMethod:  "GET",
+				ProxyRequestHeaders: map[string][]string{"Authorization": {"Bearer " + s.token(smith)}},
+				ExpectedAPIServerRequestHeaders: map[string][]string{
+					"Authorization":    {"Bearer clusterSAToken"},
+					"Impersonate-User": {"smith"},
+					"X-SSO-User":       {"username-" + smith.String()},
+				},
+				ExpectedProxyResponseStatus: http.StatusForbidden,
+				RequestPath:                 podsRequestURL("alice-private"),
+				ExpectedResponse:            "invalid workspace request: access to workspace 'alice-private' is forbidden",
+			},
+			// Given banned user eve exists
+			// And   user smith exists
+			// And   smith owns a workspace named smith-community
+			// And   smith-community is publicly visible (shared with PublicViewer)
+			// When  eve requests the list of pods in smith's workspace
+			// Then  the proxy does NOT forward the request
+			// And   the proxy rejects the call with 403 Forbidden
+			"plain http actual request as banned user to community workspace": {
+				ProxyRequestMethod:          "GET",
+				ProxyRequestHeaders:         map[string][]string{"Authorization": {"Bearer " + s.token(eve, authsupport.WithEmailClaim(eveEmail))}},
```

Comment:
probably this is a scenario more suited for the proxy test but could it be that the email claim is not set for any reason, what will/should happen in that case ? the proxy should return an error since it's a mandatory claim I would say.

Reply:
in this case the `addUserContext` middleware will fail
https://github.com/codeready-toolchain/registration-service/blob/92998a6271ae9281a9efc15c7e8aa40021f53466/pkg/proxy/proxy.go#L110

because of this check:
https://github.com/codeready-toolchain/registration-service/blob/92998a6271ae9281a9efc15c7e8aa40021f53466/pkg/auth/tokenparser.go#L80

Reply:
Right, I see that it throws an error when these claims are not being set, but I don't see a test checking this unless I'm missing something. If it's missing should we add a test case in the `proxy_community_test.go` suite which checks that if the `claims.Email` and `claims.PreferredUsername` are not set an error is returned ?

Reply:
good idea, let me work on this

Reply:
As `proxy_test.go` already contains a similar check for the JWT's `subject` field, I added a test case there: 150dbd02baf7e935b66fd56ccb22db1663b41d35